### DISCUSSION TITLE
Plugin to update my sieve filter

### DISCUSSION THREAD

**Discussion by bilogic** [the-djmaze/snappymail#1753] [Q&A]
I'm very new to snappymail, but quite amazed with it so far.

1. Is possible to write a plugin to add a button next to the sender address when viewing emails?
2. I like to classify senders as `Ads/News` and `Block`
3. This then adds that address into a sieve filter which files their messages into the respective folders

**Comment by the-djmaze** [comment]
Not at the moment.
However, there's a search filters extension (written by a user) that could do some things for you.
It's not fully the same, but it might help you for now.

**Reply by bilogic** [reply]
Hmm ok, I have a set of ideas which I feel gives a foundation layer to what I think is a new way to deal with emails. 
I'm asking because snappymail is in PHP and seems to have sufficient features to start from (instead of scratch)

Should I present them as a whole or list them 1 by 1 separately?

**Reply by the-djmaze** [reply]
Currently it is not possible to create extensions that integrate with Sieve.
This is mostly due to the complexity of Sieve.

People ask things, but Sieve has a major issue: only 1 script can be active.
Yet, it is possible to let the active script include other scripts.

However, each application prefers their own script (and disables the others).
Thunderbird, roundcube, etc always say "please activate my script".

SnappyMail doesn't care which script is active. You can edit all of them.

But when you want to automate things:
1. Which script(s) is/are active?
2. Can it be modified without other app (which owns the script) is crashing?
3. What if the rule is deleted by other application?

Yes it's terrible...

**Reply by bilogic** [reply]
Do you mean I can only edit the active script? Well, I would just edit a script identified by its name, it will be up to the user (me) to ensure I don't pick another script. What I meant is additional UI on snappy that leads to the modification happening.

**Reply by the-djmaze** [reply]
Say you also use Thunderbird (it's fictional, could be any other email app).
In Thunderbird you've added Sieve rules (in a Thunderbird script).
Now through SnappyMail you want to add a Sieve rule `if sender address then classify as Ads/News and Block`
SnappyMail finds 5 Sieve scripts:
1. USER
2. rainloop,user
3. rainloop,user.raw
4. thunderbird (active)
5. kmail

SnappyMail could add the rule to 4, but then Thunderbird might crash when it has no clue how to parse it (while it works on the Sieve server).
Instead, SnappyMail could add the rule to a new script 6 and modify 4 to include 6. If Thunderbird doesn't crash you can edit script 4 in Thunderbird, but will it save the `include`?
We potentially still have an issue!
Now SnappyMail add the rule to a new script 6 and include 4 and activates script 6. This is safe because Thunderbird script is untouched.

Now Sieve scripts:
1. USER
2. rainloop,user
3. rainloop,user.raw
4. thunderbird
5. kmail
6. snappy (active)

Now you open Thunderbird. Thunderbird sees that script 6 is active and not 4.
Thunderbird asks you to activate script 4 else the changes you make are not active (according to Thunderbird).
You say "yes" and now script 4 is active.
But... the rules you just added to 6 are now inactive!
![3a3f671d-812c-4f86-98b6-44419b904874-image-resized](https://github.com/user-attachments/assets/47b9aec5-bea7-4913-9afb-9d2f8ea304c5)


If every Sieve editor in apps just supports everything, then there is no issue.

A good example of the mess is the SnappyMail Filter editor as well.
The "Simple" script is actually the old "rainloop.user" script as a simple GUI that RainLoop created.
At the bottom it has a button 
![afbeelding](https://github.com/user-attachments/assets/7467e924-9d19-472f-97da-8728cf44f38f)
to show you the real script source.
You see the mess that RainLoop requires?

So editing this script with a different email app will break it as well.
Also read https://github.com/the-djmaze/snappymail/issues/80

It would be best if every email app uses the Kolab USER script:
```
# USER Management Script
#
# This script includes the various active sieve scripts
# it is AUTOMATICALLY GENERATED. DO NOT EDIT MANUALLY!
# 
# For more information, see http://wiki.kolab.org/KEP:14#USER
#

require ["include"];

include :personal "rainloop.user.raw";
include :personal "kmail";
```
But how is it automatically generated?
Can it be updated outside kolab?
If so, any :personal script can be (de)activated

**Reply by bilogic** [reply]
I'm not familar with sieve architecture at all, but what I'm hearing you say is: sieve script is triggered only when running the email client e.g. SnappyMail/Thunderbird.

I think email clients run in 2 forms:

1. Can reside on the IMAP server with constant connection, e.g. SnappyMail
2. Resides on standalone on one of the client desktop/laptop/mobile. e.g. Thunderbird

I'm referring to 1, and I think we can exclude many of the considerations if we setup a cron job to run the sieve filters, or even better, get the IMAP server to run the sieve filter on every new mail, and this sieve filter can be modified by SnappyMail using the fixed name approach.

Is what I proposed feasible?

**Reply by the-djmaze** [reply]
No, what i mean is:
- Sieve can have 1000 scripts
- Only 1 can be active
- Many mail apps don't understand the 1000 scripts, only the one the app created
- You are screwed when using different apps

**Reply by bilogic** [reply]
Ok, let's handle 2 parts:

1. Using only SnappyMail - any problems? I think no
2. Using SnappyMail + others
- No problems if other clients don't run sieve
- If they run sieve, they cannot crash, if they crash, it's their problem

If IMAP can run sieve, then all clients do not need sieve, even better right? 

And if SnappyMail can edit that sieve script, but not run it, instead let IMAP run it, then all clients don't have to run script (no crashes at all)

So my proposal:

1. Any email client can edit script
2. But script only executed by IMAP

**Reply by bilogic** [reply]
![image](https://github.com/user-attachments/assets/1a4812b3-8cae-47a1-a163-589b3c84275e)

@the-djmaze i'm curious, where are the scripts stored?
